# Demonstrations of server functioning

## TODO: THIS DOCUMENT NEEDS CHECKING

## Using real data
Due to their size, the real data is not included in github.  It can at present be downloaded from the below links:

[AC587 (a small dataset, < 100 samples)](https://www.dropbox.com/s/3hfgmue3g24extq/AC587.tar?dl=0)  
[AA041 (a larger dataset, ~ 1000 samples)](https://www.dropbox.com/s/l7o0v427wnu69l9/AA041.tar?dl=0)  
For large scale testing, ~ 36500 samples are available.  Please contact us if you require these.  

Download the tar files, extract them, and place them in /demos.
The files are .mfasta.gz files.  These are (i) generated by the Public Health England bioinformatics pipeline used for TB processing,
which is freely available at https://github.com/oxfordmmm/CompassCompact. (ii) with mixed bases represented by IUPAC characters, as  described [here](https://github.com/davidhwyllie/VCFMIX), (iii) gzipped. 



### AC587  
a collection of 43 mapped samples containing related TB isolates, as well as unrelated control TB samples.
The latter are added before the 43 related samples, as they are used by the server to estimate expected N frequencies in real data.
To run the demo:
- make sure mongodb is running
-- start the server  
``` pipenv run python findNeighbour4-server.py demos/AC587/config/config.json ```

or without clustering

``` pipenv run python findNeighbour4-server.py /demos/AC587/config/config_nocl.json ```  

-- run the software adding samples to the server  
``` pipenv run python demo_ac587.py ```

### AA041
a larger collection of ~ 1000 mapped samples containing related TB data.
To run the demo:
- make sure mongodb is running  
-- start the server  
``` pipenv run python findNeighbour4-server.py demos/AA041/config/config.json ```   
or (faster, without clustering)
``` pipenv run python findNeighbour4-server.py demos/AA041/config/config_nocl.json ```

-- run the software adding samples to the server  
``` pipenv run python demo_aa041.py ```

### Large scale testing
a collection of over 36,500 TB samples from diverse locations  
i) .csv files containing the positions of [where different bases map with high quality to the same site across each genome](https://github.com/davidhwyllie/VCFMIX)  
ii) fasta files in which these positions are marked with [IUPAC codes](https://www.bioinformatics.org/sms/iupac.html).  
These fasta files can be loaded by scripts similar to those above.
-- the script demo_pheall.py is suitable for doing this.

-- if not already running start the server with 
```./fn4_startup.sh demos/phe_dev/config_phe_dev_nocl.json```		
Note: (1) this config file does not do clustering; (2) it will run processes in the background using pipenv.  

-- edit demo_pheall.py to change the fastadir to the directory in which they are contained  
-- then submit the files to the server     
``` pipenv run python3 demo_pheall.py ```  

 



