##Using simulated data

This page include details of the code performing the simulations described in the
publication *M. tuberculosis microvariation is common and is associated with transmission:  analysis of three years prospective universal sequencing in England*)

The code is in two parts.  
*make_simulation.py* builds a series of phylogenies using various user-defined parameters, in the mutates these sequences using the error profiles associated with MiSeq and HiSeq sequencing machines.  It additionally introduces mixtures of sequences, whereby a new sequence is generated by combining two sequences from the phylogeny.  The positions differing between these sequences are marked as either uncertain (N) or mixed (M).
  
*run_simulation.py* loads each set of phylogenetic into the findNeighbour server, and determines whether samples contain mixed nucleotides, using the mixPORE algorithm.

